File: ports/jagentic-core/src/test/java/org/jagentic/core/OrchestrationTest.java
package org.jagentic.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpServer;

import org.junit.jupiter.api.Test;

/** Phase D — saga rollback, MoSCoW context compaction, A2A peer-as-tool (against an
 * in-process HttpServer stub). All offline/deterministic. */
class OrchestrationTest {

  private static final ObjectMapper MAPPER = new ObjectMapper();

  @Test
  void sagaRollsBackInReverseOnFailure() {
    List<String> log = new ArrayList<>();
    Saga saga = new Saga();
    saga.step("charge", () -> {
      log.add("charge");
      return "ch_1";
    }, () -> log.add("refund"));
    saga.step("ship", () -> {
      log.add("ship");
      return "sh_1";
    }, () -> log.add("cancel-ship"));

    RuntimeException boom = assertThrows(RuntimeException.class, () ->
        saga.step("reserve", () -> {
          throw new IllegalStateException("inventory gone");
        }, () -> log.add("unreserve")));
    assertEquals("inventory gone", boom.getMessage());

    // reserve's do failed (no undo recorded); ship + charge compensate in reverse
    assertEquals(List.of("charge", "ship", "cancel-ship", "refund"), log);
  }

  @Test
  void contextCompactionKeepsHighestPriorityWithinBudget() {
    List<ContextItem> items = List.of(
        new ContextItem("M".repeat(40), Priority.MUST),
        new ContextItem("S".repeat(40), Priority.SHOULD),
        new ContextItem("C".repeat(40), Priority.COULD),
        new ContextItem("W".repeat(40), Priority.WONT));
    // each item ~ (40+3)/4 = 10 tokens; budget 22 keeps two highest
    List<ContextItem> kept = new ContextWindowManager(22).compact(items);
    List<Priority> prios = kept.stream().map(ContextItem::priority).toList();
    assertTrue(prios.contains(Priority.MUST));
    assertTrue(prios.contains(Priority.SHOULD));
    assertFalse(prios.contains(Priority.COULD));
    assertFalse(prios.contains(Priority.WONT));
  }

  @Test
  void a2aClientCardSendAndPeerTool() throws IOException {
    HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
    server.createContext("/.well-known/agent-card.json", ex -> {
      byte[] body = "{\"name\":\"Stub Agent\"}".getBytes(StandardCharsets.UTF_8);
      ex.getResponseHeaders().add("Content-Type", "application/json");
      ex.sendResponseHeaders(200, body.length);
      try (OutputStream os = ex.getResponseBody()) {
        os.write(body);
      }
    });
    server.createContext("/agent", ex -> {
      Map<?, ?> req = MAPPER.readValue(ex.getRequestBody(), Map.class);
      byte[] body = MAPPER.writeValueAsBytes(
          Map.of("reply", "echo: " + req.get("text"), "ok", true));
      ex.getResponseHeaders().add("Content-Type", "application/json");
      ex.sendResponseHeaders(200, body.length);
      try (OutputStream os = ex.getResponseBody()) {
        os.write(body);
      }
    });
    server.start();
    try {
      String base = "http://127.0.0.1:" + server.getAddress().getPort();
      A2AClient client = new A2AClient(base, 2);
      assertEquals("Stub Agent", client.card().get("name"));
      assertEquals("echo: hello", client.send("c1", "hello", "alice").get("reply"));

      var tool = A2AClient.peerTool(base, 2);
      @SuppressWarnings("unchecked")
      Map<String, Object> out = (Map<String, Object>) tool.apply(
          Map.of("conversation_id", "c1", "text", "delegate this"));
      assertNotNull(out);
      assertEquals("echo: delegate this", out.get("reply"));
    } finally {
      server.stop(0);
    }
  }
}

File: ports/experimental/go/core/orchestration_test.go
package core

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestSagaRollsBackInReverseOnFailure(t *testing.T) {
	var log []string
	s := NewSaga()
	_ = s.Step("charge", func() error { log = append(log, "charge"); return nil }, func() { log = append(log, "refund") })
	_ = s.Step("ship", func() error { log = append(log, "ship"); return nil }, func() { log = append(log, "cancel-ship") })
	err := s.Step("reserve", func() error { return errFn() }, func() { log = append(log, "unreserve") })
	if err == nil {
		t.Fatal("expected step error")
	}
	// reserve do failed (no undo recorded); ship + charge undo in reverse
	want := []string{"charge", "ship", "cancel-ship", "refund"}
	if len(log) != 4 || log[2] != "cancel-ship" || log[3] != "refund" {
		t.Fatalf("log = %v, want %v", log, want)
	}
}

func errFn() error { return &simpleErr{"inventory gone"} }

type simpleErr struct{ s string }

func (e *simpleErr) Error() string { return e.s }

func TestContextCompactionMoscow(t *testing.T) {
	items := []ContextItem{
		{Text: repeat("M", 40), Priority: PriorityMust},
		{Text: repeat("S", 40), Priority: PriorityShould},
		{Text: repeat("C", 40), Priority: PriorityCould},
		{Text: repeat("W", 40), Priority: PriorityWont},
	}
	kept := NewContextWindowManager(22).Compact(items)
	var prios []Priority
	for _, it := range kept {
		prios = append(prios, it.Priority)
	}
	hasMust, hasShould, hasCould, hasWont := false, false, false, false
	for _, p := range prios {
		switch p {
		case PriorityMust:
			hasMust = true
		case PriorityShould:
			hasShould = true
		case PriorityCould:
			hasCould = true
		case PriorityWont:
			hasWont = true
		}
	}
	if !hasMust || !hasShould || hasCould || hasWont {
		t.Fatalf("kept priorities = %v (want MUST+SHOULD only)", prios)
	}
}

func repeat(s string, n int) string {
	out := ""
	for i := 0; i < n; i++ {
		out += s
	}
	return out
}

func TestA2AClientAndPeerTool(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		if r.URL.Path == "/.well-known/agent-card.json" {
			json.NewEncoder(w).Encode(map[string]any{"name": "Stub Agent"})
			return
		}
		var body map[string]string
		json.NewDecoder(r.Body).Decode(&body)
		json.NewEncoder(w).Encode(map[string]any{"reply": "echo: " + body["text"], "ok": true})
	}))
	defer srv.Close()

	client := NewA2AClient(srv.URL, 2)
	card, err := client.Card()
	if err != nil || card["name"] != "Stub Agent" {
		t.Fatalf("card = %v err = %v", card, err)
	}
	out, err := client.Send("c1", "hello", "alice")
	if err != nil || out["reply"] != "echo: hello" {
		t.Fatalf("send = %v err = %v", out, err)
	}
	tool := PeerTool(srv.URL, 2)
	res := tool(map[string]any{"conversation_id": "c1", "text": "delegate this"}).(map[string]any)
	if res["reply"] != "echo: delegate this" {
		t.Fatalf("peer tool = %v", res)
	}
}
